### ISSUE TITLE
Failed execution of ViewModelBase.SaveAsync causes EditLevel Mismatch error after second SaveAsync attempt

### ISSUE THREAD

**Issue by GreatBarrier86** [MarimerLLC/csla#1197]
Perhaps i'm doing something wrong, but I noticed that if `ViewModelBase.SaveAsync` fails after `ApplyEdit` is executed, a second attempt to call `SaveAsync` causes an EditLevel Mismatch error.

I think this is because `SaveAsync` does not set the ViewModelBase.Model back to `savable`, which means BeginEdit does not refire like it would if you were to set it manually.

At the circled code, `ApplyEdit` is called as expected. Notice in the second arrow, Model is being set to the result of `savable.SaveAsync`, which in turn, causes the Model's setter to call `BeginEdit`. However, at the third arrow (the `catch` block), Model is not being set again, so `BeginEdit` never gets called, meaning the user gets a copy of the list back at the wrong edit level.

**tl;dr;**
If the try block calls `ApplyEdit` followed by `BeginEdit` (inside Model.set), shouldn't the catch block unwind that `ApplyEdit` and put it back to its prior state (at `BeginEdit`)

![image](https://user-images.githubusercontent.com/29482393/60458333-55879400-9c0c-11e9-8856-041d4e715a6b.png)

**Version and Platform**
CSLA version: 4.9
OS: iOS, Android
Platform: Xamarin

**Comment by GreatBarrier86** [comment]
No need to rush. I was just curious. I'll wait for it to be published like usual.

**Comment by GreatBarrier86** [comment]
I just noticed this, but maybe I’m wrong. Should HookChangedEvents be outside the catch? You’d want to do this even if it’s successful, yes?

**Comment by rockfordlhotka** [comment]
OK, so I've spent a lot of hours tracking this down.

And in the process have found/fixed a couple unrelated bugs - so I guess that's good.

But ultimately I think I am trying to do something here in the viewmodel class that is outside the scope of the design of n-level undo.

Specifically, there is (built into the ancient early code) an assumption that 
1. n-level undo and cloning are separate things
1. we won't clone an object graph with active undo state

There's some reasoning behind this - most notably performance and memory optimization. There's some effort made to avoid creating massive in-memory serialized streams of data that could otherwise just be copied or maintained in memory more cheaply - _as long as we know the graph will never be "moved" in that state_.

tl;dr - an actual fix for this comes in two flavors

1. Remove the clone code in `SaveAsync` because I now think that was a big mistake on my part
1. Rethink/rework the very foundation of the n-level undo design to accommodate the idea that an object graph can be cloned (and moved?) while in an active edit state

I'm clearly leaning very much toward option 1. Very few people are building apps that use n-level undo anymore - the UI style that drove that mindset came and went for the most part. As a result, doing such a huge amount of work and testing for such little payoff seems iffy to me.

But I could be convinced otherwise I suppose.

**Comment by rockfordlhotka** [comment]
Oh, all of which is *not* to say that the viewmodel class doesn't need to be fixed - it surely does!

I'm just saying that I think the "fix" is to not clone the original object. The flow should be like all other UI scenarios:

1. ApplyEdit on the original
1. Attempt to save the original
   1. The _data portal_ actually does a clone btw!
1. In case of success
   1. Call BeginEdit on result
   1. Set Model to be result
1. In case of failure
   1. Call BeginEdit on original
   1. *OR* allow the exception to flow up so the calling code can decide what to do

And ultimately it is that very, very last point that's the key here. The viewmodel implementation is, in part, corrupted by Silverlight and the lack of async/await at the time. The way it _should_ work is based on async/await, and it should NOT be eating exceptions like it does.

Clearly THAT fix is breaking - and will occur in v5.0.

So the question is what, if anything, can/should be done in v4.x?

**Comment by GreatBarrier86** [comment]
So my question to your second point, since it’s the only one I can do away from my PC is, can you UndoEdit AFTER ApplyEdit is called? My concern is that if a failure occurs, how do you undo to the pre-apply edit state?

**Comment by rockfordlhotka** [comment]
Well the scenario we're talking about, based on the design of the n-level undo feature, is that the user has explicitly triggered a "save" action. As a result:

1. any in-process edits are applied to the object graph in memory
1. the object graph is sent to the server to be persisted
1. if that server operation succeeds the client is updated with any changes made on the server
1. if that server operation fails then client is notified via exception so the user can be notified of failure

Because the user explicitly triggered the "save" action they were clearly OK with there being "no going back".

**Comment by GreatBarrier86** [comment]
In reference to your first response, since the `DataPortal` does the clone anyway, then obviously option 1 is the best answer, it seems to me.

Eating exceptions: Certainly my preference is to throw exceptions over eating them, but the Error property does give you what you need so it's not the worst thing in the world.

Here's my concern with your response to my `ApplyEdit`/`UndoEdit` question. And perhaps i'm coming at it from the wrong perspective, so your insight would be appreciated. In my Xamarin app, I have a root list of `Diary` objects (`BusinessBase`) in a `DiaryList` (`BusinessListBase`) which is shown in my `DiariesPage`. These are shown in a Xamarin Forms `ListView`. 

1. User taps on an existing `Diary` object row in the `ListView` on the `DiariesListPage`, which has a `Comments` value of **XYZ**.
2. The app presents the `DiaryDetailPage` with the editable controls, a `Save` button, and a `Cancel` button.
3. The user makes edits the `Comments` field, changing it to **ABC**, and presses `Save`, which calls `ViewModelBase.SaveAsync`.
4. The `DataPortal` attempts to call the web service, but due to poor cell reception, the `Save` fails and a error dialog shows.
5. The user realizes they won't be able to save it because they're in the middle of nowhere, so they press `Cancel` and say they'll do it again later. This pops the `DiaryDetailPage` and puts them back at the `DiariesListPage`.
6. The `DiariesListPage's` `ListView` now shows **ABC** in the row that failed to save because `CancelEdit` had no effect.
7. The user thinks the Save worked even though it didn't and doesn't think to try again because they thought it was successful.

That's my concern. Shouldn't a `Cancel` undo any changes they tried to make?

**Comment by GreatBarrier86** [comment]
Because of what i said above, it seems better to Clone the record, call ApplyEdit on the clone, and ONLY return the clone if it was successful. That way, you retain the original that remains undoable. Like i mentioned in [#773](https://github.com/MarimerLLC/cslaforum/issues/773), I think this entire problem goes away once #1191 is resolved. Do i have it wrong?

**Comment by rockfordlhotka** [comment]
That won't work though. What you describe is the current (broken) behavior.

The reason it won't work is that the cloning process doesn't clone the state stack.

**Comment by GreatBarrier86** [comment]
Oh, i thought that's what #1191 did.  Does your fix not clone the state stack now?

**Comment by GreatBarrier86** [comment]
Closing this comment as it appears 4.11.1 seems to have resolved this problem. Thank you @rockfordlhotka!
